### PR TITLE
IDE-116 Visualize the color code in formulas

### PR DESCRIPTION
**Pull request by Jr9992**
- Visualize ColorStrings in ScriptView
- Visualize ColorStrings in FormulaEditor
- Add FormulaSpannableStringBuilder Object-Class
- Add VisualizeColorString Class
- Add VisualizeColorImageSpan Class
- Add logic to getView in FormulaBrick
- Add logic to updateTextAndCursorFromInternFormula in FormulaEditorEditText
- Add logic to refreshFormulaPreviewString in FormulaEditorFragment
- Add FormulaEditorColorVisualizeTest

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.koin.java.KoinJavaComponent
+
+class FormulaEditorColorVisualizeTest {
+
+    private val COLOR_STRING_CONVERSION_CONSTANT = 16
+
+    val projectManager: ProjectManager by KoinJavaComponent.inject(ProjectManager::class.java)
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:57]: import org.koin.java.KoinJavaComponent.inject and then just use inject(ProjectManager::class.java)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        baseActivityTestRule.launchActivity()
+    }
+
+    @Test
+    fun visualizeColorInScriptView() {
+        onView(withId(R.id.brick_set_color_edit_text)).perform(click())
+
+        setTestColorWithColorPicker(0, 0)
+
+        Espresso.pressBack()
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:83]: import it properly so you can just use pressBack()

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+        onView(withId(R.id.brick_set_color_edit_text))
+            .check(matches(ViewMatchers.withText("'#0074CD ' ")))
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:86]: import properly to just use withText(...)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    fun visualizeMultipleColorsInScriptView() {
+        onView(withId(R.id.brick_set_color_edit_text)).perform(click())
+
+        setTestColorWithColorPicker(0, 0)
+
+        onFormulaEditor().performOpenCategory(LOGIC).performSelect("or")
+
+        setTestColorWithColorPicker(2, 2)
+
+        Espresso.pressBack()
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:108]: again import properly to just call function

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+        onView(withId(R.id.brick_set_color_edit_text))
+            .check(matches(ViewMatchers.withText("'#0074CD ' or '#CA0186 ' ")))
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:111]: again import properly to just call function

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    private fun setTestColorWithColorPicker(row: Int, column: Int) {
+        onView(withId(R.id.formula_editor_keyboard_color_picker))
+            .perform(click())
+
+        ColorPickerInteractionWrapper.onColorPickerPresetButton(row, column)
+            .perform(click())
+
+        Espresso.closeSoftKeyboard()
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:177]: again import properly to just call function

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+        onView(ViewMatchers.withText(R.string.color_picker_apply))
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:179]: again import properly to just call function

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                    )
+                )
+            )
+    }
+
+    private fun setTestColorWithColorPicker(row: Int, column: Int) {
+        onView(withId(R.id.formula_editor_keyboard_color_picker))
+            .perform(click())
+
+        ColorPickerInteractionWrapper.onColorPickerPresetButton(row, column)
```
Comment (ma-zea) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorColorVisualizeTest.kt:174]: again import properly to just call function, function name alone is sufficient to know what this is about

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/FormulaBrick.java [hunk 9]
```diff
@@ -111,7 +112,11 @@ public View getView(Context context) {
 		super.getView(context);
 		for (BiMap.Entry<FormulaField, Integer> entry : brickFieldToTextViewIdMap.entrySet()) {
 			TextView formulaFieldView = view.findViewById(entry.getValue());
-			formulaFieldView.setText(getFormulaWithBrickField(entry.getKey()).clone().getTrimmedFormulaString(context));
+			String text =
+					getFormulaWithBrickField(entry.getKey()).clone().getTrimmedFormulaString(context);
+			formulaFieldView.setText(
+					FormulaSpannableStringBuilder.INSTANCE.buildSpannableFormulaString(view.getContext(), text,
```
Comment (ma-zea) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/FormulaBrick.java:118]: remove '.INSTANCE', you will see the reason in the FormulaSpannableStringBuilder!

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaSpannableStringBuilder.kt [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+import android.content.Context
+import android.text.SpannableStringBuilder
+import android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
+
+object FormulaSpannableStringBuilder {
+
+    private const val BITMAP_SIZE_MULTIPLIER = 1.25f
+
+    fun buildSpannableFormulaString(context: Context, formulaString: String, textSize: Float):
```
Comment (ma-zea) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaSpannableStringBuilder.kt:33]: add '@JvmStatic' above the method declaration, see kotlin documentation for further details!

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaEditorEditText.java [hunk 11]
```diff
@@ -350,7 +350,9 @@ public boolean redo() {
 
 	private String updateTextAndCursorFromInternFormula() {
 		String newExternFormulaString = internFormula.getExternFormulaString();
-		setText(newExternFormulaString);
+		setText(
+				FormulaSpannableStringBuilder.INSTANCE.buildSpannableFormulaString(context,
```
Comment (ma-zea) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaEditorEditText.java:354]: remove '.INSTANCE', see comment in FormulaSpannableStringBuilder

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/FormulaEditorFragment.java [hunk 12]
```diff
@@ -941,7 +943,11 @@ public void endFormulaEditor() {
 
 	public void refreshFormulaPreviewString(String newString) {
 		updateBrickView();
-		formulaBrick.getTextView(currentFormulaField).setText(newString);
+		TextView formulaFieldView = formulaBrick.getTextView(currentFormulaField);
+		formulaFieldView.setText(
+				FormulaSpannableStringBuilder.INSTANCE.buildSpannableFormulaString(this.getContext(),
```
Comment (ma-zea) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/FormulaEditorFragment.java:948]: remove '.INSTANCE'

**Review by ma-zea — CHANGES_REQUESTED** [review]
good start, there are some inconsistencies when importing certain methods and a minor mistake

**Review by edonmucaj — APPROVED** [review]

**Pull request merged**